File: Solution/WebSolution/AssessmentPortal/Microservices/CandidateAnswerAPI/Controller/CandidateAnswerController.cs
using System.Diagnostics;
using Microsoft.AspNetCore.Mvc;
using Transflower.TFLAssessment.Entities;
using Transflower.TFLAssessment.Services.Interfaces;


//Controller is now responsible to handle HTTP Requests

[ApiController]
[Route("api/candidateanswer")]
public class CandidateAnswerController : ControllerBase
{
    private readonly ILogger<CandidateAnswerController> _logger;
    private readonly ICandidateAnswerService _service;
    public CandidateAnswerController(ICandidateAnswerService service, ILogger<CandidateAnswerController> logger)
    {
        _service = service;
        _logger = logger;
    }

    // Insert candidate answers of the test .
    [HttpPost("assessmentanswers/candidates/{candidateId}")]
    public async Task<IActionResult> InsertCandidateAnswers(int candidateId, [FromBody] List<CandidateAnswer> answers)
    {
        bool status = await _service.InsertCandidateAnswers(candidateId, answers);
        if (!status)
        {
            _logger.LogError("Failed to insert candidate answers for candidate ID: {CandidateId}", candidateId);
            return BadRequest("Failed to insert candidate answers.");
        }
        _logger.LogInformation("Candidate answers inserted successfully for candidate ID: {CandidateId}", candidateId);
        return Ok(new { message = "Candidate answers inserted successfully.", status = status });
    }

    [HttpGet("assessmentanswers/candidates/{candidateId}/testId/{testId}")]
    public async Task<IActionResult> GetCandidateAnswers(int candidateId, int testId)
    {

        Console.WriteLine("GetCandidateAnswers called with candidateId: " + candidateId + " and testId: " + testId);
        List<CandidateAnswer> answers = await _service.GetCandidateAnswers(candidateId, testId);
        if (answers == null || answers.Count == 0)
        {
            _logger.LogWarning("No answers found for candidate ID: {CandidateId} and test ID: {TestId}", candidateId, testId);
            return NotFound("No answers found for the specified candidate and test.");
        }
        _logger.LogInformation("Retrieved answers for candidate ID: {CandidateId} and test ID: {TestId}", candidateId, testId);
        return Ok(answers);
    }

    //http://localhost:5299/api/candidateanswer/assessmentanswers/candidates/1/tests/1/results
    [HttpGet("assessmentanswers/candidates/{candidateId}/tests/{testId}/results")]
    public async Task<IActionResult> GetCandidateAnswerResults(int candidateId, int testId)
    {
        _logger.LogInformation(
            "GetCandidateAnswerResults called with candidateId: {cand}, testId: {t}",
            candidateId,
            testId);

        List<CandidateAnswerResult> results =
            await _service.GetCandidateAnswerResultsAsync(candidateId, testId);

        if (results == null || results.Count == 0)
        {
            _logger.LogWarning(
                "No answers found for candidate {cand} test {t}",
                candidateId,
                testId);
            return NotFound("No answers found for the specified candidate and test.");
        }
        return Ok(results);
    }

    //Get candidate and test details for the candidate answers
    [HttpGet("assessmentanswers/candidates/{candidateId}/tests/{testId}/details")]
    public async Task<IActionResult> GetCandidateTestDetails(int candidateId, int testId)
    {
        _logger.LogInformation(
            "GetCandidateTestDetails called with candidateId: {cand}, testId: {t}",
            candidateId,
            testId);

        CandidateTestDetails details =
            await _service.GetCandidateTestDetails(candidateId, testId);

        if (details == null)
        {
            _logger.LogWarning(
                "No details found for candidate {cand} test {t}",
                candidateId,
                testId);
            return NotFound("No details found for the specified candidate and test.");
        }
        return Ok(details);
    }
    
}

File: Solution/WebSolution/Java/AssessmentAPI/src/main/java/com/transflower/tflassessment/controllers/CandidateAnswerController.java
package com.transflower.tflassessment.controllers;

import java.util.List;
import java.util.concurrent.CompletableFuture;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;

import com.transflower.tflassessment.entities.CandidateAnswer;
import com.transflower.tflassessment.entities.CandidateTestDetails;
import com.transflower.tflassessment.services.CandidateAnswerService;

@RestController
@RequestMapping("api/candidateanswer")
public class CandidateAnswerController {

    @Autowired
    private CandidateAnswerService svc;

    // Insert candidate answers asynchronously
    @PostMapping("/assessmentanswers/candidates/{candidateId}")
    public CompletableFuture<Boolean> insertCandidateAnswers(
            @PathVariable("candidateId") int candidateId,
            @RequestBody List<CandidateAnswer> answers) {
        return svc.insertCandidateAnswer(candidateId, answers);
    }

    // Fetch candidate answers asynchronously
    @GetMapping("/assessmentanswers/candidates/{candidateId}/testId/{testId}")
    public CompletableFuture<List<CandidateAnswer>> getCandidateAnswer(
            @PathVariable("candidateId") int candidateId,
            @PathVariable("testId") int testId) {
        return svc.getCandidateAnswer(candidateId, testId);
    }

    // Fetch candidate test details asynchronously
    @GetMapping("/assessmentanswers/candidates/{candidateId}/tests/{testId}/details")
    public CompletableFuture<CandidateTestDetails> getCandidateTestDetails(
            @PathVariable("candidateId") int candidateId,
            @PathVariable("testId") int testId) {
        return svc.getCandidateTestDetails(candidateId, testId);
    }

    // Fetch candidate answer results asynchronously
    @GetMapping("/assessmentanswers/candidates/{candidateId}/tests/{testId}/results")
    public CompletableFuture<CompletableFuture<List<CandidateAnswer>>> getCandidateAnswerResult(
            @PathVariable int candidateId,
            @PathVariable int testId) {
        return CompletableFuture.supplyAsync(() -> svc.getCandidateAnswerResult(candidateId, testId));
    }
}
